src/client.rs

```diff
@@ -0,0 +1,267 @@
+
+impl TurnkeyApiKey {
+    pub fn new() -> TurnkeyResult<Self> {
+        dotenv::dotenv().ok();
+
+        Ok(Self {
+            private_key_hex: env::var("TURNKEY_API_PRIVATE_KEY")
+                .map_err(|e| TurnkeyError::OtherError(e.to_string()))
+                .ok(),
+            public_key_hex: env::var("TURNKEY_API_PUBLIC_KEY")
+                .map_err(|e| TurnkeyError::OtherError(e.to_string()))
+                .ok(),
+        })
+    }
+}
+
+impl Turnkey {
+    pub fn new() -> TurnkeyResult<Self> {
+        dotenv().ok();
+
+        Ok(Self {
+            base_url: env::var("TURNKEY_BASE_URL")
+                .map_err(|e| TurnkeyError::OtherError(e.to_string()))?,
+            stamper: TurnkeyApiKey::new()?,
+            client: Client::new(),
+        })
+    }
+
+    pub fn stamp(&self, stamp_input: StampInput) -> TurnkeyResult<String> {
+        let private_key_bytes = hex::decode(stamp_input.private_key_hex)
+            .map_err(|e| TurnkeyError::StampError(StampError::InvalidPrivateKeyString(e)))?;
+
+        let signing_key: SigningKey =
+            SigningKey::from_bytes(FieldBytes::from_slice(&private_key_bytes))
+                .map_err(|_| TurnkeyError::StampError(StampError::InvalidPrivateKeyBytes))?;
+
+        let sig: Signature = signing_key.sign(stamp_input.sealed_body.as_bytes());
+        let stamp = ApiStamp {
+            public_key: stamp_input.public_key_hex.clone(),
+            signature: hex::encode(sig.to_der()),
+            scheme: "SIGNATURE_SCHEME_TK_API_P256".to_string(),
+        };
+
+        let json_stamp = serde_json::to_string(&stamp).unwrap();
+
+        Ok(BASE64_URL_SAFE_NO_PAD.encode(json_stamp.as_bytes()))
+    }
+
+    fn substitute_path(uri: String, substitution: SubstitutionShape) -> TurnkeyResult<String> {
+        let mut res = uri;
+        for (key, val) in &substitution {
+            let sub = format!("{{{}}}", key);
+            let subbed_str = res.replace(&sub, val);
+            if subbed_str == res {
+                return Err(TurnkeyError::OtherError(
+                    "Substitution key was not found.".to_string(),
+                ));
+            }
+            res = subbed_str;
+        }
+
+        // TODO May be a better way to handle this, but {} are not valid uri chars
+        if res.contains('{') && res.contains('}') {
+            return Err(TurnkeyError::OtherError(
+                "Did not substitute all keys.".to_string(),
+            ));
+        }
+        Ok(res)
+    }
+    pub fn construct_url(
+        &self,
+        uri: String,
+        query: QueryShape,
+        substitution: SubstitutionShape,
+    ) -> TurnkeyResult<String> {
+        let path = Turnkey::substitute_path(uri, substitution)?;
+        // TODO: Might need to check if path starts with "/", not sure how its called
+        let mut url = format!("{}{}", self.base_url, path);
+
+        // TODO: This can all be cleaned up/simplified with Url crate if needed
```

Comment:
I'm not opposed to using a URL crate if it simplifies thing. any potential reason it'd be a bad idea?